pkg/cmd/pkg/push.go

```diff
@@ -0,0 +1,159 @@
+package pkg
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/MakeNowJust/heredoc"
+	"github.com/buildkite/cli/v3/pkg/cmd/factory"
+	"github.com/buildkite/go-buildkite/v3/buildkite"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+type pushPackageConfig struct {
+	RegistrySlug  string
+	FilePath      string
+	StdinFileName string
+}
+
+const stdinFileNameFlag = "stdin-file-name"
+
+func NewCmdPackagePush(f *factory.Factory) *cobra.Command {
+	cmd := cobra.Command{
+		Use:   "push registry-name {path/to/file | --stdin-file-name filename -}",
+		Short: "Push a new package to Buildkite Packages",
+		Long: heredoc.Doc(`
+			Push a new package to Buildkite Packages. The package can be passed as a path to a file in the second positional argument,
+			or via stdin. If passed via stdin, the filename must be provided with the --stdin-file-name flag, as Buildkite
+			Packages requires a filename for the package.`),
+		Example: heredoc.Doc(`
+			$ bk package push my-registry my-package.tar.gz
+			$ cat my-package.tar.gz | bk package push my-registry --stdin-file-name my-package.tar.gz - # Pass package via stdin, note hyphen as the argument
+		`),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cfg, err := loadAndValidateConfig(cmd.Flags(), args)
+			if err != nil {
+				return fmt.Errorf("failed to validate flags and args: %w", err)
+			}
+
+			packageName := ""
+			var from io.Reader
+			switch {
+			case cfg.FilePath != "":
+				packageName = cfg.FilePath
+
+				file, err := os.Open(cfg.FilePath)
+				if err != nil {
+					return fmt.Errorf("couldn't open file %s: %w", cfg.FilePath, err)
+				}
+
+				defer file.Close()
```

Comment:
Should this come before the error check to ensure the file is closed even in the result of an error?

Reply:
ooh yup, good call

Reply:
If `os.Open` returns an error, then `file` is (probably) nil, so `defer file.Close()` before checking `err` would probably cause a nil dereference panic.